scripts/tests/cross-test-ci-all.sh

```diff
@@ -0,0 +1,93 @@
+}
+
+function use_fed_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_FEDIMINTD_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling fed binaries for version $version..."
+    FM_FEDIMINTD_BASE_EXECUTABLE="$(nix_build_binary_for_version 'fedimintd' "$version")"
+    export FM_FEDIMINTD_BASE_EXECUTABLE
+  fi
+}
+
+function use_client_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_FEDIMINT_CLI_BASE_EXECUTABLE
+    unset FM_GATEWAY_CLI_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling client binaries for version $version..."
+    FM_FEDIMINT_CLI_BASE_EXECUTABLE="$(nix_build_binary_for_version 'fedimint-cli' "$version")"
+    export FM_FEDIMINT_CLI_BASE_EXECUTABLE
+    FM_GATEWAY_CLI_BASE_EXECUTABLE="$(nix_build_binary_for_version 'gateway-cli' "$version")"
+    export FM_GATEWAY_CLI_BASE_EXECUTABLE
+  fi
+}
+
+function use_gateway_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_GATEWAYD_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling gateway binaries for version $version..."
+    FM_GATEWAYD_BASE_EXECUTABLE="$(nix_build_binary_for_version 'gatewayd' "$version")"
+    export FM_GATEWAYD_BASE_EXECUTABLE
+  fi
+}
+
+test_results="fed_version,client_version,gateway_version,exit_code\n"
+has_failure=false
+versions+=("current")
+for fed_version in "${versions[@]}"; do
+  for client_version in "${versions[@]}"; do
+    for gateway_version in "${versions[@]}"; do
+      # test-ci-all already tests binaries running the same version, so no need to run again
+      if [[ "$fed_version" == "$client_version" && "$fed_version" == "$gateway_version" ]]; then
+        continue
+      fi
+
+      use_fed_binaries_for_version "$fed_version"
+      use_client_binaries_for_version "$client_version"
+      use_gateway_binaries_for_version "$gateway_version"
+
+      >&2 echo "========== Starting backwards-compatibility run ==========="
+      >&2 echo "fed version: $fed_version"
+      >&2 echo "client version: $client_version"
+      >&2 echo "gateway version: $gateway_version"
+
+      # continue running against other versions if there's a failure
+      set +e
+      (./scripts/tests/test-ci-all.sh)
+      exit_code=$?
+      set -e
+      test_results="$test_results$fed_version,$client_version,$gateway_version,$exit_code\n"
+      [[ "$exit_code" -gt 0 ]] && has_failure=true
+
+      >&2 echo "========== Finished backwards-compatibility run ==========="
+      >&2 echo "fed version: $fed_version"
+      >&2 echo "client version: $client_version"
+      >&2 echo "gateway version: $gateway_version"
+    done
+  done
+done
+
+>&2 echo "Backwards-compatibility tests summary:"
+echo -e "$test_results" | >&2 column -t -s ','
+# CI requires explicitly exiting with an error or success code
+[[ "$has_failure" == "true" ]] && exit 1 || exit 0
```

Comment:
Alternatively, we can temporarily remove `exit 1` so this step won't fail CI and we'll still be able to investigate/fix failures. Once master is in a good state, we can reintroduce `exit 1`.

Reply:
@elsirion I see you triggered a [rerun](https://github.com/fedimint/fedimint/actions/runs/7466845389/job/20422326326?pr=4010) of CI. I'm explicitly failing CI if the backwards compatibility tests fail, so we'll need to either address the failures or allow CI to pass prior to merging.

Reply:
@dpc and I discussed out-of-band that a sane strategy for getting this PR merged is to not return an `exit 1` if there's a backwards-compatibility failure, which allows CI to pass.

Once we investigate/fix failures we can reintroduce failing CI and preventing merging if there's a breaking change introduced.